src/rocks_compaction_scheduler.cpp

```diff
@@ -28,27 +28,167 @@
 
 #define MONGO_LOG_DEFAULT_COMPONENT ::mongo::logger::LogComponent::kStorage
 
+#include <deque>
+
 #include "mongo/platform/basic.h"
 
 #include "rocks_compaction_scheduler.h"
 
-#include <atomic>
-#include <map>
-#include <memory>
-#include <string>
-
-// for invariant()
+#include "mongo/db/client.h"
+#include "mongo/stdx/condition_variable.h"
 #include "mongo/stdx/mutex.h"
 #include "mongo/util/assert_util.h"
+#include "mongo/util/background.h"
 #include "mongo/util/log.h"
 #include "rocks_util.h"
 
+#include <rocksdb/convenience.h>
 #include <rocksdb/db.h>
 #include <rocksdb/experimental.h>
 #include <rocksdb/slice.h>
 
 namespace mongo {
 
+    class CompactionBackgroundJob : public BackgroundJob {
+    public:
+        CompactionBackgroundJob(rocksdb::DB* db);
+        virtual ~CompactionBackgroundJob();
+
+        // schedule compact range operation for execution in _compactionThread
+        Status scheduleCompactOp(const std::string& begin = std::string(), const std::string& end = std::string(),
+                                 bool rangeDropped = false, const std::function<void()>& cleanup = std::function<void()>());
+
+    private:
+        // struct with compaction operation data
+        struct CompactOp {
+            void doCompact(rocksdb::DB* db) const;
+
+            std::string _start_str;
+            std::string _end_str;
+            bool _rangeDropped;
+            std::function<void()> _cleanup;
+        };
+
+        static const char * const _name;
+
+        // BackgroundJob
+        virtual std::string name() const override { return _name; }
+        virtual void run() override;
+
+        rocksdb::DB* _db;  // not owned
+
+        bool _compactionThreadRunning = true;
+        stdx::mutex _compactionMutex;
+        stdx::condition_variable _compactionWakeUp;
+        std::deque<CompactOp> _compactionQueue;
```

Comment:
elements of `std::deque` should be `CopyAssignable`. This prevents us from using `std::deque<const CompactOp>`.
But I will change automatic `CompactOp` variable in `CompactionBackgroundJob::run` to be `const CompactOp`